Re: Added some features (mainly Android)

TimEtOff opened the pull request:
## Agenda Config Page:
On the Agenda Config Page, made the back button (and native back button/gesture on Android) actually go back one level in the files/agendas structure instead of closing the config page.
The page will still close if its the last level.
I couldn't test it on iOS nor Linux, but I don't think it can cause any problem, except just not working with iOS back gesture.

<details>
<summary><h3>Preview</h3></summary>
<img src="https://github.com/user-attachments/assets/6540116b-24e6-4014-bb00-501511beca8b" alt="Config page demo gif" height="500"/>
</details>

## Homescreen widget
Added homescreen widget compatibility for the agenda with the HomeWidget flutter plugin. 
- **<ins>Currently only available on Android</ins>** because I couln't find a Mac or setup a VM for Xcode, but once I can I'll add it to iOS too. 
- Minimum size is 2x2 on Android grid, resizable
- Widget data is refreshed everytime the agenda in-app is loaded, and last update time is displayed on the widget
- Widget changes based on system language (en/fr translations), and system theme (colors stripped from the app base color themes, but not dirrectly from the dart files as its on Android native side)
- _Tried to sync widget theme with in-app theme but couldn't make it work_
- Displays the name of the event, start and end time, and location
- Doesn't show the already finished events
- Doesn't take into account disabled days set in app settings

<details>
<summary> <h3>Preview (light theme)</h3> </summary>
<img src="https://github.com/user-attachments/assets/af10190f-8d63-4827-a326-c63ffd8b3f8b" alt="Widget small (light)" height="500"/>
<img src="https://github.com/user-attachments/assets/2dabbaeb-c0cb-4461-b1a9-4fb93ec5db8e" alt="Widget medium (light)" height="500"/>
<img src="https://github.com/user-attachments/assets/27ef701b-6e1e-4398-ae5e-e29c15ed0241" alt="Widget large (light)" height="500"/>
</details>

<details>
<summary><h3>Preview (night/dark theme)</h3> </summary>
<img src="https://github.com/user-attachments/assets/1fa7ece8-1e80-4fff-b943-492f68219fb7" alt="Widget small (night/dark)" height="500"/>
<img src="https://github.com/user-attachments/assets/01b14439-6743-45ff-bceb-e88c5f299a1a" alt="Widget medium (night/dark)" height="500"/>
<img src="https://github.com/user-attachments/assets/666a71ed-1118-47d1-bfb0-a4a942260aec" alt="Widget large (night/dark)" height="500"/>
</details>

TimEtOff committed:
fix: typos in agenda

TimEtOff committed:
feat: AngendaConfigPage go back in resources tree

TimEtOff committed:
feat: Added an Android homescreen widget for the agenda

hatch01 commented on apps/onyx/lib/screens/agenda_config/page/agenda_config_page.dart:
```diff
@@ -209,7 +224,18 @@ class AgendaConfigPage extends StatelessWidget {
+                                    pageController.page!.toInt() - 1,
+                                    duration: Res.animationDuration,
+                                    curve: Curves.easeInOut,
+                                  );
```
Why is this animate page needed if we just pop the screen just one line after?

hatch01 commented on apps/onyx/lib/screens/agenda_config/page/agenda_config_page.dart:
```diff
@@ -209,7 +224,18 @@ class AgendaConfigPage extends StatelessWidget {
                                   context.read<AgendaConfigCubit>().unSearch();
+                                } else if ((state.status == AgendaConfigStatus.loaded ||
+                                    state.status == AgendaConfigStatus.chosen) &&
+                                    state.expandedResources.isNotEmpty) {
```
I think state.expandedResources.isNotEmpty is enough, as it cannot be not empty if the agenda list is not loaded.

hatch01 commented on apps/onyx/lib/screens/agenda_config/page/agenda_config_page.dart:
```diff
@@ -60,12 +60,27 @@ class AgendaConfigPage extends StatelessWidget {
+                          pageController.page!.toInt() - 1,
+                          duration: Res.animationDuration,
+                          curve: Curves.easeInOut,
+                        );
```
This is probably not needed if we pop.

hatch01 commented on apps/onyx/lib/screens/agenda_config/page/agenda_config_page.dart:
```diff
@@ -60,12 +60,27 @@ class AgendaConfigPage extends StatelessWidget {
+                      return ;
+                    } else {
+                      if (((state.status == AgendaConfigStatus.loaded ||
+                            state.status == AgendaConfigStatus.chosen) &&
```
`state.expandedResources.isNotEmpty || state.status == AgendaConfigStatus.searchResult`
is probably enough I think

hatch01 commented on apps/onyx/lib/screens/agenda/pages/agenda_page.dart:
```diff
@@ -70,6 +71,7 @@ class AgendaPage extends StatelessWidget {
                     },
                   );
                 case AgendaStatus.ready:
+                  updateEvents(state.days, settingsState.settings);
```
This implies that the widget will only be updated when we open the app and the agenda load.
Is this wanted, or may we want to set up a work manager to regularly reload the agenda?

TimEtOff replied:
The data stored is updated only when the agenda is loaded in the app, currently with the next 14 days, but the widget itself is refreshed with this data every 30 minutes to ensure finished events are no longer displayed.
I thought an update by reloading the agenda in the app should be enough as the last update label could let the user know the displayed data may not be up to date.
Not sure about the work manager thing if you want to add it though.

hatch01 replied:
Ok thank you for the info. Sounds good to me !

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,197 @@
+
+    return '${leadingHashSign ? '#' : ''}$hexA$hexR$hexG$hexB';
+  }
+}
```
Just import: `import 'package:onyx/core/extensions/extensions_export.dart';`
The same extension is already here: `apps/onyx/lib/core/extensions/color_extension.dart`

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,197 @@
+  const AgendaHomescreenWidget({super.key});
+  @override
+  State<AgendaHomescreenWidget> createState() => _AgendaHomescreenWidgetState();
+}
```
This widget seems juste unused

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,197 @@
+  Widget build(BuildContext context) {
+    return Scaffold();
+  }
+}
```
This widget seems juste unused

TimEtOff replied:
This part (and the other "unused widget" comment) should be used to load the right app page when the widget is clicked. I didn't take the time to understand how to implement it for the moment but I wanted to let it in for future uses.

hatch01 replied:
I would prefer to drop it for this time, it is very quick to recreate one (`stful` snippet on vscode)

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,197 @@
+      .toString()
+      .padLeft(2, '0');
+  return string;
+}
```
I tested with : `return datetime.toLocal().toIso8601String();` and it works perfectly on my phone.
So maybe we can just drop this function and use toLocal().toIso8601String() in the serializer

TimEtOff replied:
Oops, yeah I switched how parsing worked on Kotlin side during development, I guess the included `toIso8601String()` is working with the current method. I'll run tests for this and the rest of the serializer to know if we can get rid of it completely.

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,197 @@
+// TODO: Replace with your App Group ID
+const String appGroupId = 'fr.onyx.lyon1';
+const String iOSWidgetName = 'AgendaWidget';
+const String androidWidgetName = 'AgendaWidget';
```
I think it would be better in res `apps/onyx/lib/core/res.dart`

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,197 @@
+
+import '../screens/settings/domain/model/settings_model.dart';
+
+// TODO: Replace with your App Group ID
```
TODO Don't make sense anymore.

TimEtOff replied:
I wasn't sure if it was actually the App Group ID on iOS (see [Flutter home screen widget codelabs](https://codelabs.developers.google.com/flutter-home-screen-widgets?hl=fr#3))

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,197 @@
+import 'package:lyon1agendaclient/lyon1agendaclient.dart';
+import 'package:onyx/core/extensions/date_extension.dart';
+
+import '../screens/settings/domain/model/settings_model.dart';
```
I used the `import package:onyx/...` in the whole project, so let's keep using this everywhere.

TimEtOff replied:
Not sure about that, every file I checked uses independent imports as well and my IDE only displays an error when adding this line, could you check or be more precise ?

hatch01 replied:
`import 'package:onyx/screens/settings/domain/model/settings_model.dart';`
Seems to work fine on my side.

TimEtOff replied:
Oop didn't understand it like that, yep I'll change it.

hatch01 reviewed (changes requested):
The first review mainly focused on the dart side.
I still want to improve the serializer, which is probably possible to completely drop.
Will look at the Kotlin part later.

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  return object;
+}
```
The theme seems not to be used anywhere, so it is probably useless.

For day and event, those are already serializable in JSON. Maybe we will have to make small adjustments on the Kotlin side or force the serializer for the date in ISO format, for example.
But this would likely be way smaller and won't use dynamic.

TimEtOff replied:
Yep, couldn't make it work on Kotlin side as Android homescreen widgets have way less features than normal Android layouts. Forgot to get rid of.

TimEtOff replied:
For day and event, I'd like to keep it like this (and in `encodeDays()`), this way days and events are in a single list and events are not nested in days, as it's easier to deal with in the Android ListView.

hatch01 replied:
To be honest, I feel like there should be a simpler method; using these dynamic and day: true/false are not very elegant.

hatch01 replied:
What about doing the data transfer like this ?
```diff
diff --git a/apps/onyx/android/app/src/main/java/fr/onyx/lyon1/AgendaWidgetService.kt b/apps/onyx/android/app/src/main/java/fr/onyx/lyon1/AgendaWidgetService.kt
index c1d08849..07d31129 100644
--- a/apps/onyx/android/app/src/main/java/fr/onyx/lyon1/AgendaWidgetService.kt
+++ b/apps/onyx/android/app/src/main/java/fr/onyx/lyon1/AgendaWidgetService.kt
@@ -6,6 +6,7 @@ import android.content.Context
 import android.widget.RemoteViews
 import es.antonborri.home_widget.HomeWidgetPlugin
 import org.json.JSONArray
+import org.json.JSONObject
 import java.time.LocalDateTime
 import java.time.format.DateTimeFormatter
 import java.time.format.TextStyle
@@ -17,99 +18,111 @@ class AgendaWidgetService : RemoteViewsService() {
     }
 }
 
+sealed class DisplayItem {
+    data class DayHeader(val date: LocalDateTime) : DisplayItem()
+    data class EventItem(
+        val name: String,
+        val start: LocalDateTime,
+        val end: LocalDateTime,
+        val location: String,
+        val description: String,
+        val teacher: String
+    ) : DisplayItem()
+}
+
 class AgendaRemoteViewsFactory(
     private val context: Context,
     intent: Intent
 ) : RemoteViewsService.RemoteViewsFactory {
 
-    private var dataSet: List<Map<String, Any>> = listOf()
+    private var displayItems: List<DisplayItem> = listOf()
 
     override fun onCreate() {}
 
-    /**
-     * Returns the same list but without past events.
-     * Keeps the last day header before the first future event.
-     */
-    fun truncateList(events: List<Map<String, Any>> ): List<Map<String, Any>> {
-        val now = LocalDateTime.now()
-        var i = -1
-        // Used to display the current day header event if the first event of the day finished
-        var lastDay = 0
-        var stop = false
-        while (!stop) {
-            i++
-            if (events[i]["day"] as Boolean) {
-                lastDay = i
-                // If the day header is after today, stop
-                if (LocalDateTime.parse(
-                        events[i]["date"] as String,
-                        DateTimeFormatter.ISO_DATE_TIME).toLocalDate()
-                        .isAfter(now.toLocalDate())) {
-                    stop = true
-                    // Prevents the day header to be displayed twice
-                    if (i < events.size - 1)
-                        i++
+    override fun onDataSetChanged() {
+        val widgetData = HomeWidgetPlugin.getData(context)
+        val daysJson = widgetData.getString("agenda_events", "[]")
+
+        android.util.Log.d("AgendaWidget", "Received days JSON: $daysJson")
+
+        try {
+            val daysArray = JSONArray(daysJson)
+            val items = mutableListOf<DisplayItem>()
+            val now = LocalDateTime.now()
+
+            for (i in 0 until daysArray.length()) {
+                val dayObj = daysArray.getJSONObject(i)
+                val dayDate = LocalDateTime.parse(dayObj.getString("date"), DateTimeFormatter.ISO_DATE_TIME)
+
+                items.add(DisplayItem.DayHeader(dayDate))
+
+                val eventsArray = dayObj.getJSONArray("events")
+                for (j in 0 until eventsArray.length()) {
+                    val eventObj = eventsArray.getJSONObject(j)
+
+                    // Parse UTC and convert to local time
+                    val startUtc = LocalDateTime.parse(eventObj.getString("start"), DateTimeFormatter.ISO_DATE_TIME)
+                    val endUtc = LocalDateTime.parse(eventObj.getString("end"), DateTimeFormatter.ISO_DATE_TIME)
+
+                    // Convert UTC to local time (assuming the events are meant to be displayed in local time)
+                    val startLocal = startUtc.atZone(java.time.ZoneOffset.UTC).withZoneSameInstant(java.time.ZoneId.systemDefault()).toLocalDateTime()
+                    val endLocal = endUtc.atZone(java.time.ZoneOffset.UTC).withZoneSameInstant(java.time.ZoneId.systemDefault()).toLocalDateTime()
+
+                    android.util.Log.d("AgendaWidget", "Event: ${eventObj.getString("name")}, ends at: $endLocal (was $endUtc UTC)")
+
+                    // Only add future events (compare with local time)
+                    if (endLocal.isAfter(now)) {
+                        android.util.Log.d("AgendaWidget", "Adding future event: ${eventObj.getString("name")}")
+                        items.add(DisplayItem.EventItem(
+                            name = eventObj.getString("name"),
+                            start = startLocal,  // Use local time
+                            end = endLocal,      // Use local time
+                            location = eventObj.optString("location", ""),
+                            description = eventObj.optString("description", ""),
+                            teacher = eventObj.optString("teacher", "")
+                        ))
+                    } else {
+                        android.util.Log.d("AgendaWidget", "Skipping past event: ${eventObj.getString("name")}")
+                    }
                 }
-            } else {
-                // If the event hasn't finished, stop
-                if (
-                    LocalDateTime.parse(
-                        events[i]["end"] as String,
-                        DateTimeFormatter.ISO_DATE_TIME
-                    )
-                        .isAfter(now)
-                )
-                    stop = true
             }
-        }
-        var newEventsList: List<Map<String, Any>> = listOf(events[lastDay])
-        newEventsList = newEventsList.plus(events.subList(i, events.size))
 
-        return (newEventsList)
-    }
+            displayItems = items
 
-    override fun onDataSetChanged() {
-        val widgetData = HomeWidgetPlugin.getData(context)
-        val events = JSONArray(
-            widgetData.getString(
-                "agenda_events",
-                "[{\"day\":true,\"date\":\"" + LocalDateTime.now().toString() + "\"},{\"day\":false,\"name\":\"No data\",\"start\":\"" + LocalDateTime.now().toString() + "\",\"end\":\"" + LocalDateTime.now().plusMinutes(10).toString() + "\",\"location\":\"Please refresh in-app\"}]"))
-            .toList()
-        dataSet = truncateList(events as List<Map<String, Any>>)
+        } catch (e: Exception) {
+            displayItems = listOf()
+        }
     }
 
-    override fun getCount() = dataSet.size
+    override fun getCount() = displayItems.size
 
     override fun getViewAt(position: Int): RemoteViews {
-        val item = dataSet[position]
-        var views: RemoteViews
-
-        // If item is a day header
-        if (item["day"] as Boolean) {
-            views = RemoteViews(context.packageName, R.layout.agenda_widget_day_item)
-
-            val date = LocalDateTime.parse(item["date"] as String, DateTimeFormatter.ISO_DATE_TIME)
-
-            views.setTextViewText(R.id.date, run {
-                date.dayOfWeek.getDisplayName(TextStyle.SHORT_STANDALONE, Locale.getDefault())
-                    .replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() } } + " " +
-                    date.dayOfMonth + " " +
-                    date.month.getDisplayName(TextStyle.SHORT_STANDALONE, Locale.getDefault()))
-        } else { // If item is an event
-            views = RemoteViews(context.packageName, R.layout.agenda_widget_event_item)
-
-            val start = LocalDateTime.parse(item["start"] as String, DateTimeFormatter.ISO_DATE_TIME)
-            val end = LocalDateTime.parse(item["end"] as String, DateTimeFormatter.ISO_DATE_TIME)
-            val location = item["location"] as String
-
-            views.setTextViewText(R.id.event_name, item["name"] as String)
-            views.setTextViewText(R.id.event_details,
-                "${start.hour.toString().padStart(2, '0')}:${start.minute.toString().padStart(2, '0')}" +
-                        " - ${end.hour.toString().padStart(2, '0')}:${end.minute.toString().padStart(2, '0')}" +
-                        "\n$location")
-        }
+        val item = displayItems[position]
 
-        return views
+        return when (item) {
+            is DisplayItem.DayHeader -> {
+                val views = RemoteViews(context.packageName, R.layout.agenda_widget_day_item)
+                val date = item.date
+
+                val dayName = date.dayOfWeek.getDisplayName(TextStyle.SHORT_STANDALONE, Locale.getDefault())
+                    .replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
+                val monthName = date.month.getDisplayName(TextStyle.SHORT_STANDALONE, Locale.getDefault())
+
+                views.setTextViewText(R.id.date, "$dayName ${date.dayOfMonth} $monthName")
+                views
+            }
+
+            is DisplayItem.EventItem -> {
+                val views = RemoteViews(context.packageName, R.layout.agenda_widget_event_item)
+
+                val timeFormat = "${item.start.hour.toString().padStart(2, '0')}:${item.start.minute.toString().padStart(2, '0')}" +
+                        " - ${item.end.hour.toString().padStart(2, '0')}:${item.end.minute.toString().padStart(2, '0')}"
+
+                views.setTextViewText(R.id.event_name, item.name)
+                views.setTextViewText(R.id.event_details, "$timeFormat\n${item.location}")
+                views
+            }
+        }
     }
 
     override fun getLoadingView() = null
@@ -118,4 +131,3 @@ class AgendaRemoteViewsFactory(
     override fun hasStableIds() = true
     override fun onDestroy() {}
 }
-
diff --git a/apps/onyx/lib/home_widget/agenda_widget.dart b/apps/onyx/lib/home_widget/agenda_widget.dart
index ad793412..c94ab2a8 100644
--- a/apps/onyx/lib/home_widget/agenda_widget.dart
+++ b/apps/onyx/lib/home_widget/agenda_widget.dart
@@ -2,62 +2,28 @@ import 'dart:convert';
 
 import 'package:home_widget/home_widget.dart';
 import 'package:lyon1agendaclient/lyon1agendaclient.dart';
-import 'package:onyx/core/extensions/date_extension.dart';
 import 'package:onyx/core/res.dart';
-
-import '../screens/settings/domain/model/settings_model.dart';
-
-dynamic serializer (dynamic object) {
-  if (object is DateTime) {
-    return object.toLocal().toIso8601String();
-  } else if (object is Event) {
-    return {
-      'day': false,
-      'location': object.location,
-      'description': object.description,
-      'teacher': object.teacher,
-      'name': object.name,
-      'start': serializer(object.start),
-      'end': serializer(object.end),
-    };
-  } else if (object is Day) {
-    return {
-      'day': true,
-      'date': serializer(object.date),
-      'events': object.events.map((e) => serializer(e)).toList(),
-    };
-  }
-
-  return object;
-}
+import 'package:onyx/screens/settings/domain/model/settings_model.dart';
 
 String encodeDays(List<Day> days, DateTime today, int daysLimit, List<int> disabledDays) {
-  List<Map<String, dynamic>> eventsList = [];
-  DateTime addingDay = today;
-  while (addingDay.isBefore(today.add(Duration(days: daysLimit)))) {
-    Day day = days.firstWhere((d) => d.date.isSameDay(addingDay),
-      orElse: () => Day(addingDay, []));
-
-    if (!disabledDays.contains(day.date.weekday)) {
-      eventsList.add({
-        'day': true,
-        'date': serializer(day.date),
-      });
-      for (var event in day.events) {
-        eventsList.add(serializer(event));
-      }
-    }
-
-    addingDay = addingDay.add(Duration(days: 1));
-  }
-  return jsonEncode(eventsList, toEncodable: serializer);
+  final eventsList = days
+      .where(
+        (day) =>
+            !day.date.isBefore(today) &&
+            !day.date.isAfter(today.add(Duration(days: daysLimit - 1))) &&
+            !disabledDays.contains(day.date.weekday) &&
+            day.events.isNotEmpty,
+      )
+      .map((day) => day.toMap()) // Encode each Day to a Map for JSON serialization
+      .toList();
+  return jsonEncode(eventsList);
 }
 
 void updateEvents(List<Day> days, SettingsModel settings) {
   DateTime now = DateTime.now();
   String encodedDays = encodeDays(days, now, 14, settings.agendaDisabledDays);
   HomeWidget.saveWidgetData<String>('agenda_events', encodedDays);
-  HomeWidget.saveWidgetData<String>('agenda_last_update', serializer(now));
+  HomeWidget.saveWidgetData<String>('agenda_last_update', now.toLocal().toIso8601String());
   HomeWidget.updateWidget(
     name: Res.agendaWidgetName,
     iOSName: Res.agendaWidgetName,
```

hatch01 commented on apps/onyx/android/app/src/main/res/values-fr/strings.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="agenda_widget_last_update">"Dernière mise à jour: "</string>
+    <string name="agenda_widget_empty_list">Pas d\'événements à venir</string>
+    <string name="add_widget">Ajouter le widget</string>
+    <string name="agenda_widget_description">Affichez vos prochains événements</string>
+    <string name="no_data">Pas de données</string>
```
It would be really cool to translate into the languages supported by Onyx (French, English, Arabic, Russian).

TimEtOff replied:
French and English were already supported, added Arabic and Russian in next commit (Translated with Deepl, might not be on point).

hatch01 commented on apps/onyx/android/build.gradle.kts:
```diff
@@ -1,3 +1,6 @@
+plugins {
+    id("org.jetbrains.kotlin.android") version "2.2.10" apply false
```
I don't know if it is needed, but that's not very handy because the Kotlin version is also in `apps/onyx/android/settings.gradle.kts`

TimEtOff replied:
I guess it was something Android Studio added by itself, removed in next commit.

hatch01 commented on apps/onyx/lib/home_widget/agenda_widget.dart:
```diff
@@ -0,0 +1,66 @@
+
+void updateEvents(List<Day> days, SettingsModel settings) {
+  DateTime now = DateTime.now();
+  String encodedDays = encodeDays(days, now, 14, settings.agendaDisabledDays);
```
It would be great to have the 14 in a Res variable, and even if we get fancy, have a parameter to change it.

hatch01 commented:
Just FYI, I force-pushed to split your commit in a more atomic way.
Don't forget to `git pull --rebase`

TimEtOff commented:
Implemented your method (I agree, it's better this way), but tweaked a few things after tests and added a Res variable to choose if we want to show empty days that are not disabled, could be a user setting for later.

I'd like to note that I got a working macOS VM so I'll probably make a new PR for the iOS widget in the future (might add user settings and other small features at the same time).

hatch01 commented on apps/onyx/android/app/src/main/java/fr/onyx/lyon1/AgendaWidget.kt:
```diff
@@ -0,0 +1,101 @@
+        list.add(value)
+    }
+    return list
+}
```
Those functions seem unused.

hatch01 commented on apps/onyx/android/app/src/main/java/fr/onyx/lyon1/AgendaWidget.kt:
```diff
@@ -0,0 +1,101 @@
+            lastUpdateText += "${lastUpdate.format(
+                DateTimeFormatter.ofPattern("dd/MM/yy HH:mm"))}"
+        else
+            lastUpdateText += context.getString(R.string.no_data)
```
```suggestion
        val lastUpdateString = widgetData.getString("agenda_last_update", null)
        var lastUpdateText = context.getString(R.string.agenda_widget_last_update)

        if (lastUpdateString != null) {
            try {
                val lastUpdate = LocalDateTime.parse(lastUpdateString, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
                lastUpdateText += lastUpdate.format(DateTimeFormatter.ofPattern("dd/MM/yy HH:mm"))
            } catch (e: Exception) {
                lastUpdateText += context.getString(R.string.no_data)
            }
        } else {
            lastUpdateText += context.getString(R.string.no_data)
        }
```

hatch01 replied:
```suggestion
        val lastUpdateString = widgetData.getString("agenda_last_update", null)
        var lastUpdateText = context.getString(R.string.agenda_widget_last_update)

        if (lastUpdateString != null) {
            try {
                val lastUpdate = LocalDateTime.parse(lastUpdateString, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
                lastUpdateText += lastUpdate.format(DateTimeFormatter.ofPattern("dd/MM/yy HH:mm"))
            } catch (e: Exception) {
                lastUpdateText += context.getString(R.string.no_data)
            }
        } else {
            lastUpdateText += context.getString(R.string.no_data)
        }
```

hatch01 reviewed (approved):
Thanks, I'll rebase and merge!
